pkg/scheduler/ugm/group_tracker.go

```diff
@@ -0,0 +1,87 @@
+/*
+ Licensed to the Apache Software Foundation (ASF) under one
+ or more contributor license agreements.  See the NOTICE file
+ distributed with this work for additional information
+ regarding copyright ownership.  The ASF licenses this file
+ to you under the Apache License, Version 2.0 (the
+ "License"); you may not use this file except in compliance
+ with the License.  You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package ugm
+
+import (
+	"sync"
+
+	"github.com/apache/yunikorn-core/pkg/common/resources"
+)
+
+type GroupTracker struct {
+	groupName    string
+	applications map[string]bool
+	queue        *QueueTracker
+
+	sync.RWMutex
+}
+
+func NewGroupTracker(group string) *GroupTracker {
+	queueTracker := NewQueueTracker("root")
+	groupTracker := &GroupTracker{
+		groupName:    group,
+		applications: make(map[string]bool),
+		queue:        queueTracker,
+	}
+	return groupTracker
+}
+
+func (gt *GroupTracker) increaseTrackedResource(queuePath, applicationID string, usage *resources.Resource) error {
+	gt.Lock()
+	defer gt.Unlock()
+	gt.applications[applicationID] = true
+	return gt.queue.increaseTrackedResource(queuePath, applicationID, usage)
+}
+
+func (gt *GroupTracker) decreaseTrackedResource(queuePath, applicationID string, usage *resources.Resource, removeApp bool) error {
+	if removeApp {
+		gt.Lock()
+		defer gt.Unlock()
+		delete(gt.applications, applicationID)
+	}
```

Comment:
In this code, we only do `gt.Lock()`, if `removeApp` is true. This is not the case with the increase pair, where we call `gt.queue.increaseTrackedResource()` within a lock. Could this be an issue? At least for consistency reasons, I think the behavior should be the same.